### DISCUSSION TITLE
`foreach` questions

### DISCUSSION THREAD

**Discussion by thedeemon** [flix/flix#6304] [General]
Does `foreach` always require an `Iterable`?

As a newcomer, two things I wanted to do and didn't know how:

1) Iterate over characters of a string. 

2) Iterate over numbers from 0 to 99, for example.

String has `iterator` method just like in `Iterable` but it's not `Iterable` because it's not a polymorphic container (kind `Type->Type`). Similarly, a range of integers `start..end`, if we represented it as a value, also wouldn't be a polymorphic container and couldn't be an `Iterable` usable in `foreach`.

I guess I can do the things above with `Iterator.range` and `Iterator.forEach`, but maybe there's a more convenient way? Many other languages do it trivially in `foreach`.

**Comment by magnus-madsen** [comment]
Excellent questions. Ill try to answer one by one:

> Does foreach always require an Iterable?

Yes, I think that is the plan.

> Iterate over characters of a string.

This is not yet possible with `foreach`. Funny you should mention it, we were just discussing it yesterday. We are currently working on associated types which will fix this issue. It will also allow iteration through e.g. `Map[k, v]`. Note that you can still use `String.iterator`, but of course without syntactic sugar that its a bit cumbersome. There is also just plain `String.map`.

> Iterate over numbers from 0 to 99, for example.

We don't have an explicit `Range` type, but I think maybe `Iterator.range` could work for you. There is also `DelayList.range` and `List.range`. In general though you would probably not want to iterate through ranges, but rather through some collection. Once we have the "new" `Iterable` in place, I think we could also add a `Range` data type to support this second scenario.

Hope that helps and let us know if you have more questions!

**Reply by thedeemon** [reply]
Thanks!

What are your thoughts on associated types vs. multi-parameter type classes? In terms of simplicity of implementation and benefits / troubles they bring for users and for language creators.

For the fun of it, I made a wrapper that turns a string or a range into an Iterable to use in `foreach`. Please don't judge the naming, I was doing it on my phone (did you know you can run Flix compiler on an Android phone in Termux? it works!).

```
enum Ible[e] {
    case MkIter(Int32, Int32 -> e)
}

instance Iterable[Ible] {
    pub def iterator(rc: Region[r], s: Ible[a]): Iterator[a, r, r] \ r =
        let Ible.MkIter(n, f) = s;
        Iterator.range(rc, 0, n) |> Iterator.map(f)
}

def ible(s:String):Ible[Char] =
    Ible.MkIter(String.length(s), i -> String.charAt(i, s))

def main() : Unit \ IO =
    let s = "abc";
    foreach(c <- ible(s))
        println(c)
```

Speaking of `Iterator[a, r, r]`, you mentioned writing a paper about it and regions/effects, is it ready yet?

**Reply by magnus-madsen** [reply]
> did you know you can run Flix compiler on an Android phone in Termux? it works!)

Wow! Neat :)

**Reply by magnus-madsen** [reply]
> Speaking of Iterator[a, r, r], you mentioned writing a paper about it and regions/effects, is it ready yet?

Not really, but https://doc.flix.dev/mutable-data.html might be sufficient.

**Reply by magnus-madsen** [reply]
> What are your thoughts on associated types vs. multi-parameter type classes? In terms of simplicity of implementation and benefits / troubles they bring for users and for language creators.

We -- like Rust -- think assoc types are more natural. But its a long discussion best had over a beer :)

**Reply by magnus-madsen** [reply]
@mlutze Check out the above code and comments, its related to what we discussed recently.
